### ISSUE TITLE
Saving & reading the optimized model

### ISSUE THREAD

**Issue by owen-deepskill** [stanfordnlp/dspy#617]
Compiling takes some time so I wouldn't want my users repeat the same process every time they run the product.
Having said that, how can I save the model and reuse it next time without re-compiling?

Specifically, in this tutorial,
https://drchrislevy.github.io/posts/dspy/dspy.html
optimized_cot_qa = teleprompter.compile(cot_qa, trainset=trainset, valset=valset)
How would I be able to save `optimized_cot_qa` so that I can reuse?

Thanks!

**Comment by insop** [comment]
You could do `save`, `load`, and use `dump_state` to check compiled states.
```
optimized_cot_qa.save("file.json")

loaded_optimized_cot_qa = CoT()
loaded_optimized_cot_qa.load("file.json")
```

**Comment by owen-deepskill** [comment]
Thanks so much @insop !

**Comment by insop** [comment]
could you do the `dump_state()` for original module and loaded module and comare?

```
loaded_optimized_cot_qa.dump_state()
```

**Comment by owen-deepskill** [comment]
I tried but still stuck with `"augmented" not in demo or not demo.augmented` and then `AttributeError: 'dict' object has no attribute 'augmented'`. Am I missing anything?

```
>>> loaded_optimized_cot_qa = CoT()
>>> loaded_optimized_cot_qa.load("optimized_cot_qa.json")
>>> loaded_optimized_cot_qa.dump_state()
>>> evaluate(loaded_optimized_cot_qa)

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dspy\evaluate\evaluate.py", line 128, in __call__                                                              | 0/126 [00:00<?, ?it/s] 
    reordered_devset, ncorrect, ntotal = self._execute_multi_thread(wrapped_program, devset, num_threads, display_progress)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dspy\evaluate\evaluate.py", line 63, in _execute_multi_thread
    example_idx, example, prediction, score = future.result()
                                              ^^^^^^^^^^^^^^^
  File "C:\Users\owenc\anaconda3\Lib\concurrent\futures\_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\owenc\anaconda3\Lib\concurrent\futures\_base.py", line 401, in __get_result
    raise self._exception
  File "C:\Users\owenc\anaconda3\Lib\concurrent\futures\thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dspy\evaluate\evaluate.py", line 116, in wrapped_program
    raise e
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dspy\evaluate\evaluate.py", line 101, in wrapped_program
    prediction = program(**example.inputs())
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dspy\primitives\program.py", line 29, in __call__
    return self.forward(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<stdin>", line 6, in forward
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dspy\predict\predict.py", line 49, in __call__
    return self.forward(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dspy\predict\chain_of_thought.py", line 59, in forward
    return super().forward(signature=signature, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dspy\predict\predict.py", line 90, in forward
    x, C = dsp.generate(template, **config)(x, stage=self.stage)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dsp\primitives\predict.py", line 77, in do_generate
    prompt = template(example)
             ^^^^^^^^^^^^^^^^^
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dsp\templates\template_v2.py", line 206, in __call__
    rdemos = [
             ^
  File "C:\Users\owenc\anaconda3\Lib\site-packages\dsp\templates\template_v2.py", line 210, in <listcomp>
    ("augmented" not in demo or not demo.augmented)
                                    ^^^^^^^^^^^^^^
AttributeError: 'dict' object has no attribute 'augmented'
```

**Comment by arnavsinghvi11** [comment]
Hi @owen-deepskill , could you share the code for compiling `optimized_cot_qa`. Ideally, the saving and loading is correct as you've done it, but I believe this error would be coming from some discrepancies in the original compiled model. 

Feel free to reference this [example](https://github.com/stanfordnlp/dspy/blob/51d2d4f9bd5f94ca624a5f0aad18a09d707a91d5/examples/nli/scone/scone.ipynb#L929) again for saving/loading compiled models.

**Comment by insop** [comment]
Hi @owen-deepskill 

What version of DSPy are you using, if you are not using the latest, then you might want to try the latest version.

**Comment by owen-deepskill** [comment]
Thanks @arnavsinghvi11  and @insop . I tested the example that @arnavsinghvi11 pointed and it worked for me, which tells me to inspect my original code. Compared to the example with ScoNt, I can spot the difference is that I used `compile` method. Am I supposed not to do this?

```
from datasets import load_dataset
import dspy

ds = load_dataset("maveriq/bigbenchhard", "penguins_in_a_table")["train"]
examples = [dspy.Example({"question": r["input"], "answer": r["target"]}).with_inputs("question") for r in ds]
trainset = examples[0:20]
valset = examples[20:]

llm = dspy.OpenAI(model="gpt-3.5-turbo-0125", max_tokens=250)
dspy.settings.configure(lm=llm, temperature=0)

import re

def eval_metric(true, prediction, trace=None):
    pred = prediction.answer
    matches = re.findall(r"\([A-Z]\)", pred)
    parsed_answer = matches[-1] if matches else ""
    return parsed_answer == true.answer

from dspy.evaluate import Evaluate

evaluate = Evaluate(devset=valset, metric=eval_metric, num_threads=6, display_progress=True, display_table=10)

from dspy.teleprompt import BootstrapFewShotWithRandomSearch

class CoT(dspy.Module):
    def __init__(self):
        super().__init__()
        self.prog = dspy.ChainOfThought("question -> answer")

    def forward(self, question):
        return self.prog(question=question)

cot_qa = CoT()

evaluate(cot_qa)

config = dict(max_bootstrapped_demos=1, max_labeled_demos=4, num_candidate_programs=4, num_threads=6)
teleprompter = BootstrapFewShotWithRandomSearch(metric=eval_metric, **config)
optimized_cot_qa = teleprompter.compile(cot_qa, trainset=trainset, valset=valset)

optimized_cot_qa.save("optimized_cot_qa.json")

loaded_optimized_cot_qa = CoT()
loaded_optimized_cot_qa.load("optimized_cot_qa.json")

evaluate(loaded_optimized_cot_qa)

```

**Comment by arnavsinghvi11** [comment]
@owen-deepskill Could you try using dspy from source instead of pypi? Seems like past [errors](https://github.com/stanfordnlp/dspy/issues/504) related to this are fixed through that but if not, we can take a closer look!

**Comment by owen-deepskill** [comment]
@arnavsinghvi11 Apologies for the delayed response. Thanks for the help. Since I encountered this problem, I changed my IDE and DSPy, etc. And for some reason I can see it is working now. Not sure what was wrong in the previous configuration. I will report if I see this error again. Many thanks!

**Comment by ajinkyaathlye** [comment]
Hey, the load function does not seem to be working. I have compiled a program using a couple of optimizers (BootstrapFewShot, BootstrapFewShotWithOptuna) and then saved it using the save method. It seems to have saved it fine (although it holds a frustratingly small amount of information after almost 15 hours of training), but when I load it all I get is a None value.

Here is how I have used it:

```
class MultiHopRAG(dspy.Module):
    # Module logic
    return dspy.Prediction()

teleprompter = BootstrapFewShotWithOptuna()
compiled_program = teleprompter.compile(MultiHopRAG)
compiled_program.save('./multi_hop_rag')
loaded_program = MultiHopRAG().load('./multi_hop_rag')
```


The `loaded_program` is simply a None value.

**Comment by ALittleBrother** [comment]
> Hey, the load function does not seem to be working. I have compiled a program using a couple of optimizers (BootstrapFewShot, BootstrapFewShotWithOptuna) and then saved it using the save method. It seems to have saved it fine (although it holds a frustratingly small amount of information after almost 15 hours of training), but when I load it all I get is a None value.嘿，加载功能似乎不起作用。我使用几个优化器（BootstrapFewShot、BootstrapFewShotWithOptuna）编译了一个程序，然后使用 save 方法保存它。它似乎保存得很好（尽管经过近 15 个小时的训练，它保存的信息量少得令人沮丧），但当我加载它时，我得到的只是 None 值。
> 
> Here is how I have used it:
> 
> ```
> class MultiHopRAG(dspy.Module):
>     # Module logic
>     return dspy.Prediction()
> 
> teleprompter = BootstrapFewShotWithOptuna()
> compiled_program = teleprompter.compile(MultiHopRAG)
> compiled_program.save('./multi_hop_rag')
> loaded_program = MultiHopRAG().load('./multi_hop_rag')
> ```
> 
> The `loaded_program` is simply a None value.

Hi, I have encountered the same problem. Have you solved it?

**Comment by ALittleBrother** [comment]
> Hey, the load function does not seem to be working. I have compiled a program using a couple of optimizers (BootstrapFewShot, BootstrapFewShotWithOptuna) and then saved it using the save method. It seems to have saved it fine (although it holds a frustratingly small amount of information after almost 15 hours of training), but when I load it all I get is a None value.
> 
> Here is how I have used it:
> 
> ```
> class MultiHopRAG(dspy.Module):
>     # Module logic
>     return dspy.Prediction()
> 
> teleprompter = BootstrapFewShotWithOptuna()
> compiled_program = teleprompter.compile(MultiHopRAG)
> compiled_program.save('./multi_hop_rag')
> loaded_program = MultiHopRAG().load('./multi_hop_rag')
> ```
> 
> The `loaded_program` is simply a None value.

The `load()` function has no return value, Maybe you should split the last step like this:
```python
loaded_program = MultiHopRAG()
lloaded_program.load('./multi_hop_rag')
# continue your logic
```

**Comment by owen-deepskill** [comment]
@fireking77 Not sure if this helps. It didn't work when I was using Python 3.11 + VSCode whereas it did with Python 3.12 + PyCharm. Can't tell what exactly fixed it but hope that gives you some clues. I haven't tested with DSPy 2.5.x as yet.

**Comment by fireking77** [comment]
> @fireking77 Not sure if this helps. It didn't work when I was using Python 3.11 + VSCode whereas it did with Python 3.12 + PyCharm. Can't tell what exactly fixed it but hope that gives you some clues. I haven't tested with DSPy 2.5.x as yet.

@owen-deepskill  thanks, but this is know fixed in 2.5.13.
